### PR TITLE
Add jsdocs and fix documentation.

### PR DESCRIPTION
**Pull request by JSAssassin**
Note: `jsdoc2md *.js > api.md` doesn't seem to be generating documents for all the `.js` files. I am not sure if I am using it incorrectly or maybe it just doesn't work.

Addresses #72

### CODE REVIEW

**Comment by mattcollier** [comment]
@JSAssassin it seems as though you may have been unaware of this outstanding PR from @mandyvenables https://github.com/digitalbazaar/edv-client/pull/71

It appears that she update the jsdocs in large swaths of code as well.  This will of course produce merge conflicts.

@JSAssassin whenever you break ground on a new PR, it's always a good idea to familiarize yourself with any open PRs on the project.  This helps to avoid duplication of effort.

**Comment by JSAssassin** [comment]
@mattcollier yeah I didn't check the open PRs, sorry about that. I'll keep that in mind and will do next time.

**Review by dmitrizagidulin — APPROVED** [review]
👍 
This is great, more docs is always good.

**Comment by aljones15** [comment]
sorry about this one @JSAssassin I guess I should have checked that there is an outstanding PR for jsdocs updates in edv client before noticing it.

**Comment by aljones15** [comment]
@mattcollier my apologies about this I noticed some issues with the jsdoc in edv client in [another PR](https://github.com/digitalbazaar/bedrock-web-profile-manager/pull/73) in `web-profile-manager`.

Those issues are still there though:

https://github.com/digitalbazaar/edv-client/blob/4a70c37890a54885f52e5f28cd11b8a4dafab9c1/EdvClient.js#L437-L458

`options.doc` has a type `{string}` for instance, but then there is a check to ensure it an object.

the same issue occurs here:

https://github.com/digitalbazaar/edv-client/blob/4a70c37890a54885f52e5f28cd11b8a4dafab9c1/EdvClient.js#L344-L376

in both cases `options.doc` is documented as a `string` when it should be an object. I might add using a string in both cases will result in an error.

@JSAssassin @dmitrizagidulin

**Comment by JSAssassin** [comment]
@aljones15 I have fixed the `options.doc` type in jsdocs. Do we need auto api document generation here? 
I tried to set it up, but jsdoc2md *.js > api.md doesn't seem to be working.

**Comment by aljones15** [comment]
> @aljones15 I have fixed the `options.doc` type in jsdocs. Do we need auto api document generation here?
> I tried to set it up, but jsdoc2md *.js > api.md doesn't seem to be working.

that has been broken for awhile. I would not worry about it unless specifically assigned to it. Thanks for the doc fix.

File: EdvClient.js [hunk 1]
```diff
@@ -63,46 +66,48 @@ export class EdvClient {
...
-   * @param {Object} options - The options to use.
+   * @param {object} options - The options to use.
    * @param {string|Array} options.attribute - The attribute name or an array of
    *   attribute names to create a unique compound index.
-   * @param {boolean} [options.unique=false] Should be `true` if the index is
+   * @param {boolean} [options.unique=false] - Should be `true` if the index is
    *   considered unique, `false` if not.
    */
   ensureIndex({attribute, unique = false}) {
-    return this.indexHelper.ensureIndex({attribute, unique});
```
Comment (aljones15) [inline EdvClient.js:73]: Are you sure we want to lose this return? I'm pretty confident it doesn't break anything, but would like a search for `ensureIndex` to ensure we are not doing something like:

```js
const foo = edvClient.ensureIndex({attribute});
if(foo) {
  bar();
}
```

File: EdvDocument.js [hunk 2]
```diff
@@ -77,20 +78,21 @@ export class EdvDocument {
   }
 
   /**
+   * @typedef Readable
```
Comment (aljones15) [inline EdvDocument.js:81]: I like these type defs, but I believe this should be `ReadableStream` and I don't believe you need the typedef unless `ReadableStream` is not imported in this file.

File: EdvDocument.js [hunk 3]
```diff
@@ -77,20 +78,21 @@ export class EdvDocument {
...
    * Encrypts and updates this document in its EDV.
    *
-   * @param {Object} options - The options to use.
-   * @param {Object} options.doc - The unencrypted document to update/insert.
-   * @param {Readable} [options.stream] a WHATWG Readable stream to read
+   * @param {object} options - The options to use.
+   * @param {object} options.doc - The unencrypted document to update/insert.
+   * @param {Readable} [options.stream] - A WHATWG Readable stream to read
```
Comment (aljones15) [inline EdvDocument.js:86]: I'm pretty confident this should be `ReadableStream`

File: EdvClient.js [hunk 4]
```diff
@@ -63,46 +66,48 @@ export class EdvClient {
...
+    this.indexHelper.ensureIndex({attribute, unique});
   }
 
   /**
+   * @typedef Readable
```
Comment (aljones15) [inline EdvClient.js:80]: this probably should be `ReadableStream`. We also might not need the typedef if `ReadableStream` is imported in this file.

File: EdvClient.js [hunk 5]
```diff
@@ -13,35 +13,38 @@ const DEFAULT_CHUNK_SIZE = 1048576;
 
 export class EdvClient {
   /**
+   * @typedef https
```
Comment (aljones15) [inline EdvClient.js:16]:
```suggestion
   *  An object that can make HTTPS requests such as node's https.Agent or ky.
   * @typedef httpsAgent
   * @see https://nodejs.org/api/https.html#https_class_https_agent
```

File: EdvClient.js [hunk 6]
```diff
@@ -13,35 +13,38 @@ const DEFAULT_CHUNK_SIZE = 1048576;
...
-   * @param {Object} [defaultHeaders=undefined] - An optional defaultHeaders
-   *   object to use with HTTP requests.
+   * @param {Function} [options.keyResolver=this.keyResolver] - A default
+   *   function that returns a Promise that resolves a key ID to a DH public
+   *   key.
+   * @param {object} [options.keyAgreementKey=null] - A default KeyAgreementKey
+   *   API for deriving shared KEKs for wrapping content encryption keys.
+   * @param {object} [options.hmac=null] - A default HMAC API for blinding
+   *   indexable attributes.
+   * @param {https.Agent} [options.httpsAgent=undefined] - An optional
```
Comment (aljones15) [inline EdvClient.js:42]: We are not importing node's `https` library and for good reason (it's backend only). So if you are ok with the change in your typedef above then replace `@param {https.Agent}` with `@param {httpsAgent}`

**Review by aljones15 — CHANGES_REQUESTED** [review]
Great work here, just need to clean up the `@typedef` issues and this should be amazing. Thanks alot.

**Review by aljones15 — APPROVED** [review]
Hey this is an improvement in docs in master so we should probably get this in. 
good work here and thanks for the updates.

**Pull request merged**